src/firefly/test/edu/caltech/ipac/astro/IpacTableReaderTest.java

```diff
@@ -0,0 +1,801 @@
+
+        //Not test the followings:
+        //dt0.getMayBeNull(); //false!!???
+        //dt0.getFormatInfo().toString(); //"edu.caltech.ipac.util.DataType$FormatInfo@...."??
+        //dt0.getImportance().toString(); //"HIGH" (who assigned this?)
+        //dt0.getMaxDataWidth(); //10
+
+        //Check the Attributes (comments):
+        List<DataGroup.Attribute> attributes = dataGroup.getKeywords();
+        if (attributes.size() == 0){
+            //System.out.println("No attributes detected.");
+            //Assert.assertTrue("No attributes", noAttributes);
+            Assert.assertFalse("No attributes found", !noAttributes);
+        }
+        else {
+            int commentNum = 0;
+            for (DataGroup.Attribute att : attributes) {
+                if (att.isComment()) {
+                    //System.out.println(att.toString());
+                    Assert.assertEquals("check the comments:", att.toString(), attributeComments[commentNum]);
+                    commentNum++;
+                }
+            }
+        }
+
+        //Check the attributes (key, value):
+        Map<String, DataGroup.Attribute> attributeMap = dataGroup.getAttributes();
+        //attributeMap.size();
+        int j = 0;
+        String value;
+        for (Map.Entry entry : attributeMap.entrySet()) {
+            Assert.assertEquals("check the key", entry.getKey(), attributeKeys[j]);
+            value = ((DataGroup.Attribute) entry.getValue()).getValue().toString();
+            //System.out.println(value + " " + attributeValues[j] + " " + j);
+            Assert.assertEquals("check the value", value, attributeValues[j]);
+            j++;
+        }
+
+        //Check the data content/values:
+        List<DataObject> objList = dataGroup.values(); //
+        for (int row = 0; row < objList.size(); row++) {
+            for (int col = 0; col < dataTypes.length; col++) {
+                String dataExpected = objList.get(row).getDataElement(dataTypes[col]).toString();
+                Assert.assertEquals("check the 0th data value in the row", dataExpected, dataValues[row][col]);
+            }
+        }
+    }
+
+
+    // All the methods:
+
+    //public static DataGroup readIpacTable(File f, String catName)
+    //public static DataGroup readIpacTable(File f, String onlyColumns[], boolean useFloatsForDoubles, String catName)
+    //public static DataGroup readIpacTable(File f, String onlyColumns[], boolean useFloatsForDoubles, String catName, boolean isHeadersOnlyAllow)
+
+    //public static DataGroup readIpacTable(Reader fr, String catName)
+    //public static DataGroup readIpacTable(Reader fr, String catName, long estFileSize)
+    //public static DataGroup readIpacTable(Reader fr, String catName, String onlyColumns[],boolean useFloatsForDoubles,long estFileSize)
+    //public static DataGroup readIpacTable(Reader fr, String catName, String onlyColumns[],boolean useFloatsForDoubles,long estFileSize, boolean isHeadersOnlyAllow)
+
+
+}
```

Comment:
This unit test serves more like a verification of the algorithm and less like a unit test.  However since building a small DataGroup (table) is very easy,  the unit test can really be written as a unit test.  

For unit test:
1. There is a one to one correspondence for each public method in the main class.  
For example, there should be:
    testIpacTableReaderTwoArgs()
    testIpacTableReaderThreeArgs()
    testIpacTableReaderFourArgs()
    testIpacTableNoDataException()

2. Unit test does not check if the algorithm is correct, it checks if the algorithm does what it should do. 
    For example, no matter what table passed to the reader, if the reader reads it correctly, the  test is pass.

3. If there is duplicated columns, it is the issue in TableDef. If the TableDef allows to have duplicated columns, it is OK, if it does not allow duplication, TableDef should check it.

4.  The no header is also the issue with the TableDef.  It should catch such issue. 

Overall, I think this test can serve as a wonderful use cases verifications.  It correctly identify a few holes in the current algorithm such duplication in the column names, and empty headers etc.  GREAT Job!

To make it a self-contained unit test,
   1. build needed dataGroup dynamically.  The built dataGoup will be the expected DataGroup.  
   2. Let IpacTableReader reads the dataGroup from the output file by the the same dataGroup and do the comparison.
  3. In the tearDown method:
   remove the temporary output files used for the test. 

Then the unit test is self-contain, neat and clean.